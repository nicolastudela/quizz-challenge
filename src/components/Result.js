import React from "react";
import { Typography } from "@material-ui/core";
import PropTypes from "prop-types";
import { Redirect } from "react-router-dom";
import { connect } from "react-redux";

export const Result = ({ userName, location, results }) => {
  const result = results.length > 0 ? results[results.length - 1] : null;

  return result ? (
    <>
      <Typography
        variant="h3"
        align="center"
      >{`${userName} you scored `}</Typography>
      <Typography
        variant="h4"
        align="center"
      >{`${result.overall.answersCorrect} / ${result.overall.totalQuestions}`}</Typography>
    </>
  ) : (
    <Redirect
      to={{
        pathname: "/",
        state: { from: location }
      }}
    />
  );
};

Result.propTypes = {
  userName: PropTypes.string.isRequired,
  location: PropTypes.shape({
    state: PropTypes.object
  }).isRequired,
  results: PropTypes.arrayOf(
    PropTypes.shape({
      overall: PropTypes.object
    })
  )
};

Result.defaultProps = {
  results: []
};

const mapStateToProps = state => {
  return {
    results: state.user.results
  };
};

export default connect(mapStateToProps)(Result);
