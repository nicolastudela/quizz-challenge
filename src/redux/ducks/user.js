// FILL ACTIONS CREATORS AND EXPORT THEM

const CREATE = "user/CREATE";
const ADD_RESULT = "user/ADD_RESULT";

const createUser = createdUser => ({ type: CREATE, payload: { createdUser } });

export const userSignIn = name => {
  const user = { name, results: [] };
  localStorage.setItem("user", JSON.stringify(user));
  return createUser(user);
};

export const addResult = result => {
  const user = JSON.parse(localStorage.getItem("user"));
  user.results.push(result);
  localStorage.setItem("user", JSON.stringify(user));
  return { type: ADD_RESULT, payload: result };
};

const user = JSON.parse(localStorage.getItem("user"));
const initialState = user || {};

export default (state = initialState, action) => {
  switch (action.type) {
    // do reducer stuff
    case CREATE: {
      const { createdUser } = action.payload;
      return { ...state, ...createdUser };
    }
    case ADD_RESULT: {
      const result = action.payload;
      return { ...state, results: state.results.concat([result]) };
    }
    default:
      return state;
  }
};
