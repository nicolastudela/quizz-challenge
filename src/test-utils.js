/* eslint-disable react/prop-types */
import React from "react";
import { ThemeProvider } from "@material-ui/core/styles";
import { MemoryRouter } from "react-router-dom";
import { Provider } from "react-redux";
import thunk from "redux-thunk";
import configureStore from "redux-mock-store";
import theme from "./theme";

export const INITIAL_EMPTY_STORE = {
  user: {},
  questionnaire: {
    items: [],
    isFetching: false,
    error: null
  }
};

const mockStore = configureStore([thunk]);

export const BasicTestWrapper = ({ children }) => {
  return (
    <ThemeProvider theme={theme}>
      <MemoryRouter>{children}</MemoryRouter>
    </ThemeProvider>
  );
};

export const WithReduxStoreTestWrapper = ({
  children,
  store = INITIAL_EMPTY_STORE,
  initialEntries,
  initialIndex
}) => {
  return (
    <ThemeProvider theme={theme}>
      <MemoryRouter initialEntries={initialEntries} initialIndex={initialIndex}>
        <Provider store={mockStore(store)}>{children}</Provider>
      </MemoryRouter>
    </ThemeProvider>
  );
};
