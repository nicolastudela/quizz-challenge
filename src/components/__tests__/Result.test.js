import React from "react";
import renderer from "react-test-renderer";
import { userWith1BooleanResult } from "../../__mocks__/user.mock";
import { BasicTestWrapper } from "test-utils";
import { Result } from "../Result";

jest.mock("react-router-dom", () => {
  const original = jest.requireActual("react-router-dom");
  return {
    ...original,
    Redirect: jest.fn().mockImplementation(() => <div> REDIRECT</div>)
  };
});

describe("Result page", () => {
  it("renders propery when results are available", () => {
    expect.assertions(1);
    const component = renderer.create(
      <BasicTestWrapper>
        <Result results={userWith1BooleanResult.results} />
      </BasicTestWrapper>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("renders propery when results are NOT available", () => {
    expect.assertions(1);
    const component = renderer.create(
      <BasicTestWrapper>
        <Result results={[]} />
      </BasicTestWrapper>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
