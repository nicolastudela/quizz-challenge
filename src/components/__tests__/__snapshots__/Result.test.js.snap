// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Result page renders propery when results are NOT available 1`] = `
<div>
   REDIRECT
</div>
`;

exports[`Result page renders propery when results are available 1`] = `
Array [
  <h3
    className="MuiTypography-root MuiTypography-h3 MuiTypography-alignCenter"
  >
    You scored
  </h3>,
  <h4
    className="MuiTypography-root MuiTypography-h4 MuiTypography-alignCenter"
  >
    2 / 3
  </h4>,
  <div
    className="MuiBox-root MuiBox-root-32"
  />,
  <ul
    className="MuiList-root MuiList-padding"
  >
    <li
      className="MuiListItem-root MuiListItem-gutters"
      disabled={false}
    >
      <div
        className="MuiListItemIcon-root"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root"
          focusable="false"
          role="presentation"
          style={
            Object {
              "color": "green",
            }
          }
          viewBox="0 0 24 24"
        >
          <path
            d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
          />
        </svg>
      </div>
      <div
        className="MuiListItemText-root"
      >
        <span
          className="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
        >
          QUEST 1: Question 1
        </span>
      </div>
    </li>
    <li
      className="MuiListItem-root MuiListItem-gutters"
      disabled={false}
    >
      <div
        className="MuiListItemIcon-root"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root"
          focusable="false"
          role="presentation"
          style={
            Object {
              "color": "red",
            }
          }
          viewBox="0 0 24 24"
        >
          <path
            d="M19 13H5v-2h14v2z"
          />
        </svg>
      </div>
      <div
        className="MuiListItemText-root"
      >
        <span
          className="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
        >
          QUEST 1: Question 2
        </span>
      </div>
    </li>
    <li
      className="MuiListItem-root MuiListItem-gutters"
      disabled={false}
    >
      <div
        className="MuiListItemIcon-root"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root"
          focusable="false"
          role="presentation"
          style={
            Object {
              "color": "green",
            }
          }
          viewBox="0 0 24 24"
        >
          <path
            d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
          />
        </svg>
      </div>
      <div
        className="MuiListItemText-root"
      >
        <span
          className="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
        >
          QUEST 1: Question 3
        </span>
      </div>
    </li>
  </ul>,
  <div
    className="MuiBox-root MuiBox-root-65"
  />,
  <div
    className="MuiBox-root MuiBox-root-66"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "width": "30%",
        }
      }
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        Other Results
      </span>
    </button>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "width": "50%",
        }
      }
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        Play Again?
      </span>
    </button>
  </div>,
]
`;
